### PR TITLE
sql: replace the test for #14473 with a better one

### PR DESCRIPTION
**Pull request by andreimatei**
14473 was a panic when preparing a statement, under certain conditions.

- add a targeted test for #14473
  - The previous commit that added the test was unclear about the
  conditions that it meant to test. It also had too many statements
  without explanation of which one is expected to crash and why. It was
  also affecting other unrelated subtests by introducing an unnecessary
  testing knob. Part of the problem was that it needed to test a very
  particular set of conditions, and it was doing that from the wrong
  level (pgwire vs sql).

- revert a funky testing knob: the knob caused the LeaseStore to always
acquire a lease when queried by id, but not when queried by name. Also
there are other ways to remove leases without the need for this knob.

Here's some details about the exact conditions necessary for the panic in 14473:

- The panic occurs when the LeaseStore attempts to create a fake planner
to execute internal sql statements. If the client.Txn object doesn't
have the OrigTimestamp field set -> panic. Preparing statements didn't
use to set that field.

- In most cases, however, attempting to acquire a table lease is done by
using a table name. The process of resolving the name performs a KV Get,
which means that the OrigTimestamp is initialized by the TxnCoordSender.

- So, in order to crash we need to (also) acquire some leases by id.
This is done, e.g., by edits on tables with FKs (they reference the
parents by ID).

- In conclusion, we need to be in a particular situation where the
prepare doesn't need to resolve any names (all tables referenced by name
have a lease in the cache), but also try to reference by ID a table
that's not in the cache.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14641)

### CODE REVIEW

**Comment by andreimatei** [comment]
@nvanbenschoten how are we looking on having the client always initialize that `txn.OrigTimestamp`? The fact that the txn sometimes initializes it internally is confusing :)

**Comment by vivekmenezes** [comment]
I think you need to discuss why the particular flow of events cannot be recreated without all these testing knobs.
The test you have created is certainly more complex than what it is replacing.

I think the prepare tests are all in pgwire just because we can only call prepare as part of pgwire today.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/executor.go, line 300 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Kh-kQeXugZMJ8iGvrVx:-Kh-kQeXugZMJ8iGvrVy:b-lhcha) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b1810962a566738e63c000e942660b3e4de89e6/pkg/sql/executor.go#L300)):*
> ```Go
> 	// BeforePrepare is called by the Executor before preparing any statement. It
> 	// gives access to the planner that will be used to do the prepare. If any of
> 	// the return values are not nil, the values are used as the prepare results
> ```

There is only one return value here.

I'd call this GetPreparedStatement 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-:-Kh-m7UkkK_dOF8ebGgD:b-rus8i7)*

**Comment by andreimatei** [comment]
I actually had written the test originally for reproducing the bug with just sql:
https://github.com/cockroachdb/cockroach/compare/master...andreimatei:prepare-lease-test-alt?expand=1
You can take a look and see if you like it :)

But then I decided that it's not the way to go - we don't care about reproducing a very contrived set of conditions; they're very fragile. Any number of small changes in how we prepare edits on tables with FKs would have changed the set of conditions necessary. 

What we want to test is extremely simple - that we can acquire a lease (from the LeaseStore) inside a prepare (technically we want to test that we can create an internal planner, but acquiring a lease is an easier to understand and more stable requirement). SQL is just the wrong level at which to test this. Don't be put off by boilerplate; the test I've written is extremely simple:

```
SQLExecutor: &ExecutorTestingKnobs{
			BeforePrepare: ... {
    _, err := s.LeaseManager().(*LeaseManager).Acquire(
					ctx, planner.txn, sqlbase.ID(dummyTableID), 0 /* version */)
    return nil, errors.New("a-OK")
})

_, err = db.PrepareContext(context.TODO(), query)
if !testutils.IsError(err, "a-OK") {
	t.Fatal(err)
}
```

That's it. Executing a `LeaseManager.Acquire` in the context that the Executor has setup for a prepare. Super simple. No particular setup is required. The bulk of the code is not necessary for reproducing the bug; it's asserting that we tested what we thought we were testing (the previous test did not do that).

The test before was in the wrong place - the fact that we can't execute prepares as a statement doesn't mean that any test involving preparing needs to be in the pgwire package. The issue being tested had nothing to do with pgwire (the protocol and network components). A gazillion of our tests execute statements through the pgwire interface; it doesn't mean they're pgwire tests.

Also, the test was failing on a call to `db.Exec`, not on a call to `db.Prepare`, for extra confusion. Also, it was extremely unclear what it was testing. It had 3 statements - 2 inserts, one update. It's very hard to understand why only the 3rd one fails. A naive (or even not so naive) reader would expect the first one to fail. After you understand why that doesn't fail, you expect the 2nd one to fail. After you understand that, you're left wondering why does the 3rd one fail exactly. I had to keep notes to understand why that is.
Supremely, the test was very fragile. Very subtle changes in the executions of those statements would make the test not test what it thought it was testing.

To even get the test to fail without the fix, it needed a testing knob that was surprising, inconsistent (it disabled a very particular code path in the LeaseManager, but not other similar code paths), and not really necessary.
The knob was also set for all the subtests in that test, which is unnecessary and perhaps damaging. And generally, that test bigger test in which the new "subtest" was added is just bad. It groups a ton of sub-tests together (but you can't immediately separate them into subtests because some of the statements inadvertently depend on the others), and also it both executes statements and prepares statements. Perhaps it's OK as a general test of our implementation of the protocol, where you just want to throw random things on the wire and see that it works, but not for a targeted test (let alone a super-duper-targeted test, which this new one was trying to be).

The old knob was intrusive and unnecessary.
The knobs I've added are simple event handlers - very benign. The more of these knobs we have, the better. (Well, in my test, I am using an event handler to inject code execution, not just to record the event, but that's an established pattern and a pretty good one in my opinion). The point is, they don't change behavior in subtle and unexpected ways.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/executor.go, line 300 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Kh-kQeXugZMJ8iGvrVx:-Kh3_n0uk46hbC6TlzzF:b4x8rx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b1810962a566738e63c000e942660b3e4de89e6/pkg/sql/executor.go#L300)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

There is only one return value here.

I'd call this GetPreparedStatement 
</blockquote></details>

This is an event handler (and `BeforePrepare` is an event-handler-like name). The fact that it can optionally return something is secondary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-:-Kh-pcTDI4urOL92L5ZF:bpg744z)*

**Comment by jordanlewis** [comment]
:lgtm: Thank you very much for following up with this - the test you added is indeed much more targeted and precise.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-:-KhxPMFaarvX_jC8kol-:b78y34q)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor_test.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-KhytQAxXI1D_sLQMiFd:-KhytQAyQLcn5CBpcSkT:b-7s4sqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L60)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	query := "SELECT 1"
> ```

const

---

*[pkg/sql/executor_test.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-KhytbyFri18ESvEG4eP:-KhytbyFri18ESvEG4eQ:b-u2twpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L74)):*
> ```Go
> 				ctx context.Context, stmt string, planner *planner,
> 			) (*PreparedStatement, error) {
> 				if query != stmt {
> ```

can this happen?

---

*[pkg/sql/executor_test.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Khyr5DGnUIo-GhzYzbg:-Khyr5DGnUIo-GhzYzbh:bkusyb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L123)):*
> ```Go
> 
> 	// Create a dummy table so we have something to get a lease on.
> 	_, err := db.Exec("CREATE DATABASE d;" +
> ```

don't really need the +

---

*[pkg/sql/executor_test.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Khys4Tk-m7SWhDuQyqq:-Khys4Tk-m7SWhDuQyqr:b-vve27a) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L129)):*
> ```Go
> 	}
> 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "d", "t")
> 	atomic.StoreUint32(&dummyTableID, uint32(tableDesc.ID))
> ```

Can we create the table right at the top, read the table descriptor and use that id. I think then you don't need the exp == 0 condition above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641)*

**Comment by vivekmenezes** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-:-KhyuMPMv8jNiVri2C9u:bnfp4nl)*

**Comment by andreimatei** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/executor_test.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-KhytbyFri18ESvEG4eP:-Kj5E5iChUqiRWR72kYT:b-hli8up) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L74)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

can this happen?
</blockquote></details>

I think so... This test runs multiple statements. And even if it didn't, I'm not sure whether internal SQL queries run through the Executor or not. Even if they don't today, I think they might in the future. I think it's generally in the spirit of these filters to ignore what they're not concerned with.

---

*[pkg/sql/executor_test.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Khyr5DGnUIo-GhzYzbg:-Kj5ElGSIx2HyHVCgj_8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L123)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

don't really need the +
</blockquote></details>

Done.

---

*[pkg/sql/executor_test.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-Khys4Tk-m7SWhDuQyqq:-Kj5FUw2q-sWwx_3dDEz:b-roj7cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L129)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can we create the table right at the top, read the table descriptor and use that id. I think then you don't need the exp == 0 condition above
</blockquote></details>

well it's a chicken and egg thing - we can't create tables until we've created the cluster, and creating the cluster needs the knob filters. We could install shim filters that allow us to dynamically install other filters (in fact, I think there's even some utilities standardizing such a shim for some filter, but not the ones used here), but I think that'd result in code for not enough reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641)*

**Comment by andreimatei** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/executor_test.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-KhytQAxXI1D_sLQMiFd:-Kj5GT1SPE5EvV45bYKz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/296145d402697de62defb07a938257ef5c8cf550/pkg/sql/executor_test.go#L60)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

const
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641)*

**Comment by andreimatei** [comment]
TFTR

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14641#-:-Kj5LqgfZgK_z2up5wKp:bb73zsl)*
